aiida/orm/querybuilder.py

```diff
@@ -989,20 +993,36 @@ def _get_aiida_entity_res(value) -> Any:
         except TypeError:
             return value
 
-    def first(self) -> Optional[List[Any]]:
-        """Executes the query, asking for the first row of results.
+    @overload
+    def first(self, flat: Literal[False]) -> Optional[list[Any]]:
+        ...
+
+    @overload
+    def first(self, flat: Literal[True]) -> Optional[Any]:
+        ...
+
+    def first(self, flat: bool = False) -> Optional[list[Any] | Any]:
+        """Return the first result of the query.
 
-        Note, this may change if several rows are valid for the query,
-        as persistent ordering is not guaranteed unless explicitly specified.
+        Calling ``first`` results in an execution of the underlying query.
 
+        Note, this may change if several rows are valid for the query, as persistent ordering is not guaranteed unless
+        explicitly specified.
+
+        :param flat: if True, return just the projected quantity if there is just a single projection.
         :returns: One row of results as a list, or None if no result returned.
         """
         result = self._impl.first(self.as_dict())
 
         if result is None:
             return None
```

Comment:
@sphuber it can return `None` here

Reply:
Sure, I saw that, but take the first hit `aiida/orm/nodes/data/upf.py:136:27:`
```python
        existing_upf = builder.first()

        if existing_upf is None:
            ...
        else:
            existing_upf = existing_upf[0]
```
That looks like a false positive to me. If anything this would be the perfect candidate for `first(flat=True)`.

Reply:
Same goes for the next "hits"
```
tests/orm/test_querybuilder.py:724:26: E1136: Value 'result' is unsubscriptable (unsubscriptable-object)
tests/orm/test_querybuilder.py:725:26: E1136: Value 'result' is unsubscriptable (unsubscriptable-object)
```
Also false positives.

Reply:
Fair, then IMO, I would just disable `unsubscriptable-object`, because pylint does not seem to be doing a good job with them: https://github.com/PyCQA/pylint/issues/1498
I don't feel pylint should be trying to involve itself in type-checking, since that is what mypy is for, that does it a lot better

Reply:
Eurgh could also be overload, could be future annotations: https://github.com/PyCQA/pylint/issues/5189, https://github.com/PyCQA/pylint/issues/3979, https://github.com/PyCQA/pylint/issues/4369
Silly pylint